File: leetcode/226_invert_binary_tree.py
"""
T: O(N)
S: O(N)

Just do as the description says.
"""


class TreeNode:

    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:

    def invertTree(self, root: TreeNode) -> TreeNode:
        nodes_to_process = [root]
        while nodes_to_process:
            curr = nodes_to_process.pop()
            if curr:
                curr.left, curr.right = curr.right, curr.left
                nodes_to_process.append(curr.left)
                nodes_to_process.append(curr.right)
        return root

File: leetcode/java/226InvertBinaryTree.java
/**
 * T: O(N) S: O(N)
 *
 * <p>Just do as the description says.
 */
import java.util.Deque;
import java.util.LinkedList;

public class TreeNode {
  int val;
  TreeNode left;
  TreeNode right;

  TreeNode() {}

  TreeNode(int val) {
    this.val = val;
  }

  TreeNode(int val, TreeNode left, TreeNode right) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

class Solution {
  public TreeNode invertTree(TreeNode root) {
    Deque<TreeNode> nodesToProcess = new LinkedList<>();
    nodesToProcess.add(root);
    while (!nodesToProcess.isEmpty()) {
      TreeNode currentNode = nodesToProcess.pop();
      if (currentNode == null) {
        continue;
      }
      TreeNode temp = currentNode.left;
      currentNode.left = currentNode.right;
      currentNode.right = temp;
      nodesToProcess.push(currentNode.left);
      nodesToProcess.push(currentNode.right);
    }
    return root;
  }
}
